### PR TITLE
Add OpenVINO backend for torch.compile node

### PR DESCRIPTION
**Pull request by openvino-dev-samples**
# Features description 

- To support both `.safetensor` model and LoRa weights with OpenVINO runtime.
- To support ALL of Intel CPU/GPU/NPU.

image

# Installation 
To enable this integration, you only need to install openvino runtime in advance:
```
pip install openvino
python3 main.py --cpu --use-pytorch-cross-attention
```

# Test cases

> * [x]  model without lora node -> lora node enabled -> lora node disabled
      # current issue: /

> * [x]  model with lora node -> lora node disabled -> lora node enabled
      # current issue: it will need 2 warm-up model compilation (generation is slow) before fast generation with LoRA node

# Screenshots

TorchCompileModel output has to be connected with the input of  KSampler. 
image

The following model has to be selected for checkpoint 
image

https://github.com/comfyanonymous/ComfyUI/issues/2473

### CODE REVIEW

**Comment by simonlui** [comment]
Questions, as an Intel Arc owner and having contributed to the repository.

1.) I have used both the Triton(`inductor`) and OpenVINO backends using custom nodes and Triton is faster in both compilation time and speed from what I have tested. How do these backends differ and plan to be supported because the benefit to me for adding OpenVINO seems minimal outside of supporting NPU devices.

2.) Are there sensible errors that pop up if you don't have a suitable OpenVINO GPU or NPU and try to run with this node and ways to diagnose how to solve them if users run into them? This can be a issue and both device types, but especially NPU, require drivers at this time to function properly and I can't even get my LNL laptop to use NPU at this time on Linux so I also have questions about maturity too at this time.

**Comment by openvino-dev-samples** [comment]
> Questions, as an Intel Arc owner and having contributed to the repository.
> 
> 1.) I have used both the Triton(`inductor`) and OpenVINO backends using custom nodes and Triton is faster in both compilation time and speed from what I have tested. How do these backends differ and plan to be supported because the benefit to me for adding OpenVINO seems minimal outside of supporting NPU devices.
> 
> 2.) Are there sensible errors that pop up if you don't have a suitable OpenVINO GPU or NPU and try to run with this node and ways to diagnose how to solve them if users run into them? This can be a issue and both device types, but especially NPU, require drivers at this time to function properly and I can't even get my LNL laptop to use NPU at this time on Linux so I also have questions about maturity too at this time.

hi @simonlui thanks for your feedback. 
1) i believe the benefit of adding openvino is to trigger Intel GPU and NPU on inferencing.

**Comment by Panchovix** [comment]
Hi there, just passed by and wanted to say, many many thanks! Finally with your add_patches modifications (and others) managed to make loras work with torch.compile. Really appreciated!

**Comment by Panchovix** [comment]
Sorry for double post, but wondering, does loading a lora, then disabling it, and then enabling it again works fine for you?

Maybe some unpatching or recompiling is needed?

I think on a first inference with a lora, it will patch the keys before compiling, and it will work.

If you then disable it and enable the lora, it will compile without a lora and will add some _orig_mod prefixes to the keys, so when trying to apply the lora keys to the compiled model, it will not match.

Correct me if I'm wrong though.

**Comment by openvino-dev-samples** [comment]
> python3 main.py --cpu --use-pytorch-cross-attention

I think it can supp

> Sorry for double post, but wondering, does loading a lora, then disabling it, and then enabling it again works fine for you?
> 
> Maybe some unpatching or recompiling is needed?
> 
> I think on a first inference with a lora, it will patch the keys before compiling, and it will work.
> 
> If you then disable it and enable the lora, it will compile without a lora and will add some _orig_mod prefixes to the keys, so when trying to apply the lora keys again on a 3rd inference to the compiled model, it will not match the key and it won't load.
> 
> Correct me if I'm wrong though.

Hi, when your implementation path start from a checkpoint without LoRa, everything works. However if it starts from a checkpoint with LoRa, the patch does not work. which mean:

- [x] model without lora -> lora enabled -> lora disabled
- [ ] model with lora -> lora disabled -> lora enabled

In second case, my new patch will not be triggered, so I believe it is a general issue for torch.compile node, and I will do furthe investigation.

**Comment by openvino-dev-samples** [comment]
> > python3 main.py --cpu --use-pytorch-cross-attention
> 
> I think it can supp
> 
> > Sorry for double post, but wondering, does loading a lora, then disabling it, and then enabling it again works fine for you?
> > Maybe some unpatching or recompiling is needed?
> > I think on a first inference with a lora, it will patch the keys before compiling, and it will work.
> > If you then disable it and enable the lora, it will compile without a lora and will add some _orig_mod prefixes to the keys, so when trying to apply the lora keys again on a 3rd inference to the compiled model, it will not match the key and it won't load.
> > Correct me if I'm wrong though.
> 
> Hi, when your implementation path start from a checkpoint without LoRa, everything works. However if it starts from a checkpoint with LoRa, the enabling and disabling LoRA does not work. which mean:
> 
> * [x]  model without lora -> lora enabled -> lora disabled
> * [ ]  model with lora -> lora disabled -> lora enabled
> 
> In second case, my new patch will not be triggered, so I believe it is a general issue for torch.compile node, and I will do furthe investigation.

I have updated the PR, however it may need 2 warm-up inference for first time generation with LoRA weights

**Comment by openvino-dev-samples** [comment]
hi @comfyanonymous could you help to review ?

**Review by NineMeowICT — APPROVED** [review]

**Comment by openvino-dev-samples** [comment]
Hi @NineMeowICT Could you help to check if this PR is ready to be merged ? thanks

**Comment by NineMeowICT** [comment]
@openvino-dev-samples I'm sorry for the misunderstanding. I'm not one of ComfyUI's contributors and I don't have write access, my approval was simply to recognize your efforts.

**Comment by raymondlo84** [comment]
How can we get this merged?

**Comment by Sen-sou** [comment]
@openvino-dev-samples Appreciate you working on implementing this, currently using your fork, it dosent seem to be using the gpu for the inference even tho ive selected gpu as the device. Is there any specefic openvino version that i need to use? 
Thank you

**Comment by openvino-dev-samples** [comment]
> @openvino-dev-samples Appreciate you working on implementing this, currently using your fork, it dosent seem to be using the gpu for the inference even tho ive selected gpu as the device. Is there any specefic openvino version that i need to use? I dont think its compiling anything and just strainght skipping the block. Thank you

hi Could you share the screenshot of network ? BTW you can try following command to update the version of openvino:
`pip install --pre -U openvino --extra-index-url https://storage.openvinotoolkit.org/simple/wheels/nightly`

**Comment by Sen-sou** [comment]
> > @openvino-dev-samples Appreciate you working on implementing this, currently using your fork, it dosent seem to be using the gpu for the inference even tho ive selected gpu as the device. Is there any specefic openvino version that i need to use? I dont think its compiling anything and just strainght skipping the block. Thank you
> 
> hi Could you share the screenshot of network ? BTW you can try following command to update the version of openvino: `pip install --pre -U openvino --extra-index-url https://storage.openvinotoolkit.org/simple/wheels/nightly`

Thanks for replying,
Updated the openvino version, it worked for once, it used the gpu for the inference, but after that it switched back to cpu because i changed the sampler. Restarted couple of times, still using the cpu. Do i have to trigger a recompile for the gpu to work? 

Heres the screenshot you asked.
workflow

**Comment by openvino-dev-samples** [comment]
> > > @openvino-dev-samples Appreciate you working on implementing this, currently using your fork, it dosent seem to be using the gpu for the inference even tho ive selected gpu as the device. Is there any specefic openvino version that i need to use? I dont think its compiling anything and just strainght skipping the block. Thank you
> > 
> > 
> > hi Could you share the screenshot of network ? BTW you can try following command to update the version of openvino: `pip install --pre -U openvino --extra-index-url https://storage.openvinotoolkit.org/simple/wheels/nightly`
> 
> Thanks for replying, Updated the openvino version, it worked for once, it used the gpu for the inference, but after that it switched back to cpu because i changed the sampler. Restarted couple of times, still using the cpu. Do i have to trigger a recompile for the gpu to work?
> 
> Heres the screenshot you asked. workflow
> 
> Thank you.

i dont think change on Sampler will lead recompiling for GPU. but I will try to reproduce your case. thanks for sharing.

**Comment by Sen-sou** [comment]
> > > > @openvino-dev-samples Appreciate you working on implementing this, currently using your fork, it dosent seem to be using the gpu for the inference even tho ive selected gpu as the device. Is there any specefic openvino version that i need to use? I dont think its compiling anything and just strainght skipping the block. Thank you
> > > 
> > > 
> > > hi Could you share the screenshot of network ? BTW you can try following command to update the version of openvino: `pip install --pre -U openvino --extra-index-url https://storage.openvinotoolkit.org/simple/wheels/nightly`
> > 
> > 
> > Thanks for replying, Updated the openvino version, it worked for once, it used the gpu for the inference, but after that it switched back to cpu because i changed the sampler. Restarted couple of times, still using the cpu. Do i have to trigger a recompile for the gpu to work?
> > Heres the screenshot you asked. workflow
> > Thank you.
> 
> i dont think change on Sampler will lead recompiling for GPU. but I will try to reproduce your case. thanks for sharing.



> > > > @openvino-dev-samples Appreciate you working on implementing this, currently using your fork, it dosent seem to be using the gpu for the inference even tho ive selected gpu as the device. Is there any specefic openvino version that i need to use? I dont think its compiling anything and just strainght skipping the block. Thank you
> > > 
> > > 
> > > hi Could you share the screenshot of network ? BTW you can try following command to update the version of openvino: `pip install --pre -U openvino --extra-index-url https://storage.openvinotoolkit.org/simple/wheels/nightly`
> > 
> > 
> > Thanks for replying, Updated the openvino version, it worked for once, it used the gpu for the inference, but after that it switched back to cpu because i changed the sampler. Restarted couple of times, still using the cpu. Do i have to trigger a recompile for the gpu to work?
> > Heres the screenshot you asked. workflow
> > Thank you.
> 
> i dont think change on Sampler will lead recompiling for GPU. but I will try to reproduce your case. thanks for sharing.

Thank you for looking, but i resolved the problem. It was the low memory which was causing it to go back to cpu, Increasing memory helped resolve this. Thanks a lot anyway.

**Comment by Kosinkadink** [comment]
Hey, sorry that this did not get an official review for so long!

For the lora fix for torch.compile, even with the keys workaround, there was indeed a fundamental issue with the way torch.compile was implemented in ComfyUI with the object_patches, with anything relating to keys still being broken while the model was loaded. I created a PR that just got merged that fixes this properly: https://github.com/comfyanonymous/ComfyUI/pull/8213, with torch.compile now being managed by a '''set_torch_compile_wrapper''' function in ```comfy_api.torch_helpers``` and needing no workarounds in ModelPatcher code. I also exposed a bunch of torch.compile kwargs on that new ```set_torch_compile_wrapper``` function, so that an openvino implementation can pass in the 'options' param.

For the OpenVino portion of the PR, managing the openvino dependency should be done in some sort of standardized manner - this can be done the cleanest by implementing the Torch Compile OpenVino node in a custom node pack. This will ensure that any time a workflow uses the OpenVino node, users are able to acquire the dependencies via ComfyUI-Manager. Creating/publishing the custom node pack is something that can be done on your end, so that you will be free to make any edits in the future.

Here are the docs for publishing a custom node to registry: https://docs.comfy.org/registry/publishing#publishing-nodes

Since the review was so delayed, I went ahead and wrote the core of the custom node code that would achieve the goal of this PR, feel free to use it/reference it:

```
import torch
import openvino as ov
import openvino.frontend.pytorch.torchdynamo.execute as ov_ex

from comfy_api.torch_helpers import set_torch_compile_wrapper


class TorchCompileModelOpenVINO:
    @classmethod
    def INPUT_TYPES(s):
        core = ov.Core()
        available_devices = core.available_devices

        return {
            "required": {
                "model": ("MODEL",),
                "backend": (["openvino"],),
                "openvino_device": (available_devices,),
            },
        }

    RETURN_TYPES = ("MODEL",)
    FUNCTION = "patch"

    CATEGORY = "_for_testing"
    EXPERIMENTAL = True

    def patch(self, model, backend, openvino_device):
        print(model.__class__.__name__)
        if backend == "openvino":
            options = {"device": openvino_device}
            torch._dynamo.reset()
            ov_ex.compiled_cache.clear()
            ov_ex.req_cache.clear()
            ov_ex.partitioned_modules.clear()
        else:
            options = None
        m = model.clone()
        set_torch_compile_wrapper(m,
                                  backend=backend,
                                  options=options,
                                  )
        return (m,)

# The node ID in NODE_CLASS_MAPPINGS should be globally unique across ComfyUI ecosystem
NODE_CLASS_MAPPINGS = {
    "OpenVINO_TorchCompileModel": TorchCompileModelOpenVINO,
}
```

Let me know if you have any questions or concerns, I'll do my best to answer promptly!

**Comment by openvino-dev-samples** [comment]
> Hey, sorry that this did not get an official review for so long!
> 
> For the lora fix for torch.compile, even with the keys workaround, there was indeed a fundamental issue with the way torch.compile was implemented in ComfyUI with the object_patches, with anything relating to keys still being broken while the model was loaded. I created a PR that just got merged that fixes this properly: #8213, with torch.compile now being managed by a `set_torch_compile_wrapper` function in `comfy_api.torch_helpers` and needing no workarounds in ModelPatcher code. I also exposed a bunch of torch.compile kwargs on that new `set_torch_compile_wrapper` function, so that an openvino implementation can pass in the 'options' param.
> 
> For the OpenVino portion of the PR, managing the openvino dependency should be done in some sort of standardized manner - this can be done the cleanest by implementing the Torch Compile OpenVino node in a custom node pack. This will ensure that any time a workflow uses the OpenVino node, users are able to acquire the dependencies via ComfyUI-Manager. Creating/publishing the custom node pack is something that can be done on your end, so that you will be free to make any edits in the future.
> 
> Here are the docs for publishing a custom node to registry: https://docs.comfy.org/registry/publishing#publishing-nodes
> 
> Since the review was so delayed, I went ahead and wrote the core of the custom node code that would achieve the goal of this PR, feel free to use it/reference it:
> 
> ```
> import torch
> import openvino as ov
> import openvino.frontend.pytorch.torchdynamo.execute as ov_ex
> 
> from comfy_api.torch_helpers import set_torch_compile_wrapper
> 
> 
> class TorchCompileModelOpenVINO:
>     @classmethod
>     def INPUT_TYPES(s):
>         core = ov.Core()
>         available_devices = core.available_devices
> 
>         return {
>             "required": {
>                 "model": ("MODEL",),
>                 "backend": (["openvino"],),
>                 "openvino_device": (available_devices,),
>             },
>         }
> 
>     RETURN_TYPES = ("MODEL",)
>     FUNCTION = "patch"
> 
>     CATEGORY = "_for_testing"
>     EXPERIMENTAL = True
> 
>     def patch(self, model, backend, openvino_device):
>         print(model.__class__.__name__)
>         if backend == "openvino":
>             options = {"device": openvino_device}
>             torch._dynamo.reset()
>             ov_ex.compiled_cache.clear()
>             ov_ex.req_cache.clear()
>             ov_ex.partitioned_modules.clear()
>         else:
>             options = None
>         m = model.clone()
>         set_torch_compile_wrapper(m,
>                                   backend=backend,
>                                   options=options,
>                                   )
>         return (m,)
> 
> # The node ID in NODE_CLASS_MAPPINGS should be globally unique across ComfyUI ecosystem
> NODE_CLASS_MAPPINGS = {
>     "OpenVINO_TorchCompileModel": TorchCompileModelOpenVINO,
> }
> ```
> 
> Let me know if you have any questions or concerns, I'll do my best to answer promptly!

Thanks for your suggestion. I will test and publish it through ComfyUI-Manager.

**Comment by openvino-dev-samples** [comment]
> Hey, sorry that this did not get an official review for so long!
> 
> For the lora fix for torch.compile, even with the keys workaround, there was indeed a fundamental issue with the way torch.compile was implemented in ComfyUI with the object_patches, with anything relating to keys still being broken while the model was loaded. I created a PR that just got merged that fixes this properly: #8213, with torch.compile now being managed by a `set_torch_compile_wrapper` function in `comfy_api.torch_helpers` and needing no workarounds in ModelPatcher code. I also exposed a bunch of torch.compile kwargs on that new `set_torch_compile_wrapper` function, so that an openvino implementation can pass in the 'options' param.
> 
> For the OpenVino portion of the PR, managing the openvino dependency should be done in some sort of standardized manner - this can be done the cleanest by implementing the Torch Compile OpenVino node in a custom node pack. This will ensure that any time a workflow uses the OpenVino node, users are able to acquire the dependencies via ComfyUI-Manager. Creating/publishing the custom node pack is something that can be done on your end, so that you will be free to make any edits in the future.
> 
> Here are the docs for publishing a custom node to registry: https://docs.comfy.org/registry/publishing#publishing-nodes
> 
> Since the review was so delayed, I went ahead and wrote the core of the custom node code that would achieve the goal of this PR, feel free to use it/reference it:
> 
> ```
> import torch
> import openvino as ov
> import openvino.frontend.pytorch.torchdynamo.execute as ov_ex
> 
> from comfy_api.torch_helpers import set_torch_compile_wrapper
> 
> 
> class TorchCompileModelOpenVINO:
>     @classmethod
>     def INPUT_TYPES(s):
>         core = ov.Core()
>         available_devices = core.available_devices
> 
>         return {
>             "required": {
>                 "model": ("MODEL",),
>                 "backend": (["openvino"],),
>                 "openvino_device": (available_devices,),
>             },
>         }
> 
>     RETURN_TYPES = ("MODEL",)
>     FUNCTION = "patch"
> 
>     CATEGORY = "_for_testing"
>     EXPERIMENTAL = True
> 
>     def patch(self, model, backend, openvino_device):
>         print(model.__class__.__name__)
>         if backend == "openvino":
>             options = {"device": openvino_device}
>             torch._dynamo.reset()
>             ov_ex.compiled_cache.clear()
>             ov_ex.req_cache.clear()
>             ov_ex.partitioned_modules.clear()
>         else:
>             options = None
>         m = model.clone()
>         set_torch_compile_wrapper(m,
>                                   backend=backend,
>                                   options=options,
>                                   )
>         return (m,)
> 
> # The node ID in NODE_CLASS_MAPPINGS should be globally unique across ComfyUI ecosystem
> NODE_CLASS_MAPPINGS = {
>     "OpenVINO_TorchCompileModel": TorchCompileModelOpenVINO,
> }
> ```
> 
> Let me know if you have any questions or concerns, I'll do my best to answer promptly!

Hi @Kosinkadink do you have tutorial on how to develop the custom node in personal repository ? I only find a tutorial for custom node in [ComfyUI/custom_nodes](https://docs.comfy.org/custom-nodes/walkthrough)
